### DISCUSSION TITLE
Stream API?

### DISCUSSION THREAD

**Discussion by DmitrySoshnikov** [dropbox/fast_rsync#8] [Q&A]
I've two questions:

- Does this library support streaming API? Example: need to calculate deltas for large files (4GB+). From the examples I found all the delta, and signatures are memory vectors
- Is this library compatible with the original `librsync` in C++?

In general would be great having more usage examples in this repo. Thanks!

**Comment by goffrie** [comment]
Sorry, I completely missed this issue!
> Does this library support streaming API? Example: need to calculate deltas for large files (4GB+). From the examples I found all the delta, and signatures are memory vectors

It doesn't. This was a choice to make the library simpler, as in our use case we only ever need to operate on buffers that are already in-memory.

> Is this library compatible with the original librsync in C++?

It isn't. `librsync` has a much more powerful and complex API that this library wouldn't be able to emulate.

**Comment by cr1901** [comment]
How practical is a simple per-block streaming API where the user (me) does most of the work in their application?

I find the idea of a library simpler and faster than `librsync` appealing, but the lack of streaming (compressing large, mostly append-only files a la incremental backups) means this library is a no-go for me for my use case.

**Reply by goffrie** [reply]
I'm not exactly sure what you mean by "per-block streaming". I'm not inherently opposed to adding a streaming API (though I don't think it makes sense for signatures since the entire thing is going to be in an in-memory hashmap); however it would be difficult to preserve the simplicity/performance of the implementation for the all-in-memory use case.

**Comment by cr1901** [comment]
>I'm not exactly sure what you mean by "per-block streaming".

This was meant to be a "one night project". Unfortunately, it became clear that as the night went on that it would _not_ be a one-night project. I think it's best to wait until I can show a code example of what I want w/ the `librsync` bindings. Unfortunately, they don't work on my platform right now (https://github.com/mbrt/librsync-rs/issues/14). So I put the project aside until I have more time than "all the deps need to just work for me to get this done". Then I can look at the linked issue and make a small working sample.

>though I don't think it makes sense for signatures since the entire thing is going to be in an in-memory hashmap

It's fine if the entire signature is in memory at once, as long as I don't have to have the entire base file in memory at once (which is sometimes larger than the amount of RAM on this machine).
